File: src/lib/inventory/rollTags.ts
import { getChampionCounterForFrame, type ChampionType } from "@/data/rules/championCounters";
import type { RollTag } from "@/lib/db/types";
import type { WeaponRecord } from "@/lib/manifest/types/records";

export interface ComputeRollTagsOptions {
  isCrafted?: boolean;
}

const CHAMPION_TAG: Record<ChampionType, RollTag> = {
  Barrier: "ChampionBarrier",
  Overload: "ChampionOverload",
  Unstoppable: "ChampionUnstoppable",
};

const PERK_CHAMPION_PATTERNS: { pattern: RegExp; tag: RollTag }[] = [
  { pattern: /\banti-?barrier\b/i, tag: "ChampionBarrier" },
  { pattern: /\bbarrier\b/i, tag: "ChampionBarrier" },
  { pattern: /\boverload\b/i, tag: "ChampionOverload" },
  { pattern: /\bunstoppable\b/i, tag: "ChampionUnstoppable" },
];

function perkNamesFromHashes(
  plugHashes: number[],
  perkNameMap: Map<number, string>,
): string[] {
  return plugHashes
    .map((hash) => perkNameMap.get(hash))
    .filter((name): name is string => Boolean(name));
}

function hasPerk(names: string[], target: string): boolean {
  const needle = target.toLowerCase();
  return names.some((name) => name.toLowerCase().includes(needle));
}

function championTagFromPerks(names: string[]): RollTag | null {
  for (const { pattern, tag } of PERK_CHAMPION_PATTERNS) {
    if (names.some((name) => pattern.test(name))) return tag;
  }
  return null;
}

/**
 * Rule-based roll tag assignment from plug hashes and optional weapon metadata.
 * Tags are computed at sync time and stored on inventory rows.
 */
export function computeRollTags(
  plugHashes: number[],
  perkNameMap: Map<number, string>,
  weaponRecord?: WeaponRecord | null,
  options?: ComputeRollTagsOptions,
): RollTag[] {
  const tags = new Set<RollTag>();
  const perkNames = perkNamesFromHashes(plugHashes, perkNameMap);

  if (options?.isCrafted) {
    tags.add("Crafted");
  }

  if (weaponRecord) {
    const intrinsic = getChampionCounterForFrame(
      weaponRecord.frame,
      weaponRecord.itemTypeName,
    );
    if (intrinsic) tags.add(CHAMPION_TAG[intrinsic]);
  }

  const perkChampion = championTagFromPerks(perkNames);
  if (perkChampion) tags.add(perkChampion);

  if (
    weaponRecord?.itemTypeName === "Hand Cannon"
    && hasPerk(perkNames, "Pugilist")
    && hasPerk(perkNames, "Swashbuckler")
  ) {
    tags.add("MeleeBuildCandidate");
  }

  if (hasPerk(perkNames, "Demolitionist") && hasPerk(perkNames, "Adrenaline Junkie")) {
    tags.add("OrbitBuild");
  }

  return [...tags];
}

File: flutter/packages/bungie/lib/src/inventory/roll_tags.dart
import 'package:destiny2_sandbox_data/destiny2_sandbox_data.dart';

/// Wire values for inventory `rollTags` (product `RollTag` union).
abstract final class RollTags {
  static const crafted = 'Crafted';
  static const championBarrier = 'ChampionBarrier';
  static const championOverload = 'ChampionOverload';
  static const championUnstoppable = 'ChampionUnstoppable';
  static const meleeBuildCandidate = 'MeleeBuildCandidate';
  static const orbitBuild = 'OrbitBuild';
}

/// Legendary weapon metadata needed for frame-based champion tags.
///
/// Subset of product `WeaponRecord` (`frame` + `itemTypeName`).
class RollTagWeaponMeta {
  const RollTagWeaponMeta({
    required this.frame,
    required this.itemTypeName,
  });

  final String frame;
  final String itemTypeName;
}

class ComputeRollTagsOptions {
  const ComputeRollTagsOptions({this.isCrafted = false});

  final bool isCrafted;
}

final _perkChampionPatterns = <({RegExp pattern, String tag})>[
  (pattern: RegExp(r'\banti-?barrier\b', caseSensitive: false), tag: RollTags.championBarrier),
  (pattern: RegExp(r'\bbarrier\b', caseSensitive: false), tag: RollTags.championBarrier),
  (pattern: RegExp(r'\boverload\b', caseSensitive: false), tag: RollTags.championOverload),
  (pattern: RegExp(r'\bunstoppable\b', caseSensitive: false), tag: RollTags.championUnstoppable),
];

String? _championTagForType(ChampionType type) {
  switch (type) {
    case ChampionType.barrier:
      return RollTags.championBarrier;
    case ChampionType.overload:
      return RollTags.championOverload;
    case ChampionType.unstoppable:
      return RollTags.championUnstoppable;
  }
}

List<String> _perkNamesFromHashes(
  List<int> plugHashes,
  Map<int, String> perkNameMap,
) {
  final names = <String>[];
  for (final hash in plugHashes) {
    final name = perkNameMap[hash];
    if (name != null && name.isNotEmpty) names.add(name);
  }
  return names;
}

bool _hasPerk(List<String> names, String target) {
  final needle = target.toLowerCase();
  return names.any((name) => name.toLowerCase().contains(needle));
}

String? _championTagFromPerks(List<String> names) {
  for (final entry in _perkChampionPatterns) {
    if (names.any((name) => entry.pattern.hasMatch(name))) {
      return entry.tag;
    }
  }
  return null;
}

/// Rule-based roll tag assignment from plug hashes and optional weapon metadata.
///
/// Port of product `computeRollTags` in `src/lib/inventory/rollTags.ts`.
/// Tags are computed at sync time and stored on inventory rows.
/// Soft metadata only — never auto-applies build edits.
List<String> computeRollTags(
  List<int> plugHashes,
  Map<int, String> perkNameMap, {
  RollTagWeaponMeta? weapon,
  ComputeRollTagsOptions? options,
  bool? isCrafted,
}) {
  final crafted = isCrafted ?? options?.isCrafted ?? false;
  final tags = <String>{};
  final perkNames = _perkNamesFromHashes(plugHashes, perkNameMap);

  if (crafted) {
    tags.add(RollTags.crafted);
  }

  if (weapon != null) {
    final intrinsic = getChampionCounterForFrame(
      weapon.frame,
      weapon.itemTypeName,
    );
    if (intrinsic != null) {
      final tag = _championTagForType(intrinsic);
      if (tag != null) tags.add(tag);
    }
  }

  final perkChampion = _championTagFromPerks(perkNames);
  if (perkChampion != null) {
    tags.add(perkChampion);
  }

  if (weapon?.itemTypeName == 'Hand Cannon' &&
      _hasPerk(perkNames, 'Pugilist') &&
      _hasPerk(perkNames, 'Swashbuckler')) {
    tags.add(RollTags.meleeBuildCandidate);
  }

  if (_hasPerk(perkNames, 'Demolitionist') &&
      _hasPerk(perkNames, 'Adrenaline Junkie')) {
    tags.add(RollTags.orbitBuild);
  }

  return tags.toList(growable: false);
}
